### ISSUE TITLE
Slow HDD read.

### ISSUE THREAD

**Issue by lelik107** [BLAKE3-team/BLAKE3#390]
Hello guys, BLAKE3 Team! I found out that any b3sum_windows_x64_bin.exe reads a HDD slow. What exactly I did:
I calculated the sums of Windows 10 iso images, 5.3 Gb and 5.8 Gb files, WD RE 1 Tb was used, I measured with a simple stopwatch.The speed of this HDD is 140-80 Mbyte/s. And it's not about a disk cache, because I rebooted after every test.
So v1.5.1, v1.3.3  ~ 5.3 Gb - 1 m 45 s,  5.8 Gb ~ 1 m 57 s;  v1.0.0 a bit faster ~ 5.3 Gb - 1 m 22 s,  5.8 Gb ~ 1 m 31 s. 
And I hear the disk spinning.
b2sum-bin_20130305.zip, 7-zip h command with any algo ~ 5.3 Gb - 45-48 s,  5.8 Gb ~ 55 s -1 m;.
And I don't hear the disk spinning.
So my simple question is:  why b3sum is this slower?

**Comment by lelik107** [comment]
@lulcat Yes, the disk cache reading is OK with Windows b3sum. IDK exactly the languages used but there might be Assembly as well. Most of the third party Windows SW, for example
https://github.com/OV2/RapidCRC-Unicode and Total Commander
use single-threaded reference C code because there’s no multi-threaded C code yet from the developers. As for a raw reading from HDD in RapidCRC Unicode and TC it shows just the same speed as any other algorithm. If you have 210 mb/s you’ll get it. For a buffered reading single-threaded C code gives 1500-1700 mb/s with all SIMDs but AVX-512F (I don’t have them). And, yes, I will be nice to have multithreaded C code, maybe OMP.

**Comment by lulcat** [comment]
Hi again. Let me correct myself. I just checked now.. .the actual RUST binary as well, which does mmap AND MT...runs AT 60 MB/s in an example case I have....

I then evict the file from cache and run ANY Other hashsum, which will beat it... e.g. b2sum? (can be sha*sum, anysum, all will run at the max ingress speed of the file)! e.g.

time b2sum same_example => 210MB/s over 3 times faster!!

I meant the C code won't do any better because it isn't optimised like rust (no mmapping and no MT , i.e. OMP).

THIS Is the issue you notice and reported I think. and it is a 'edge case' of b3 which I haven't figure dout yet WHY happens.

b2sum as you see does NOT suffer from this..  and I DO have avx512, I can run it with that in C etc but this problem persists.

it is on _slow_ spinners and _big files_ which trigger it. fast IO will not spike this case.

Now, again, as I said.. I have solved this in my environment with something on top.. but requires heuristically determining if we have that 'edge case'. I'd prefer obviously it to be solved in the b3 code but I am not sure where to touch it. I don't wanna touch crypto anyway with my edits. :)

**Comment by lelik107** [comment]
@lulcat  I'm glad you've find a solution, but I neither a developer nor a coder myself rather a end user in Windows, we just don't build SW very often :)

**Comment by lelik107** [comment]
@lulcat Yes, --no-mmap solves the issue but as I see : "Disable memory mapping. ### Currently this also disables multithreading."
It's for Rust, because C isn't multithreaded anyway. But what's the point then of using Rust and Rayon if at the and of the day we have exactly what we have in C, and I should say in more understandable form.

**Comment by oconnor663** [comment]
Apologies for not commenting here sooner. See this older issue: https://github.com/BLAKE3-team/BLAKE3/issues/31

The problem isn't really specific to BLAKE3 itself, but to the "divide and conquer" multithreading strategy we currently use. When `b3sum` hashes a file, by default you get a worker thread for every core on your machine, and each of these threads starts working on a different part of the file. From the disk's perspective, this requires a lot of seeking, and for spinning disks / HDDs in particular, it's this seeking that's slow. See also the docs for [`Hasher::update_mmap_rayon`](https://docs.rs/blake3/latest/blake3/struct.Hasher.html#method.update_mmap_rayon).

The next tricky part is that this performance downside _doesn't_ apply to SDDs or to files cached in RAM. In those cases seeking is cheap or free, and it's common to see a 5-10x performance boost from multithreading. So if we changed the default, the majority of our users would run an order of magnitude slower, and they'd probably never know why. That 5-10x number depends entirely on the number of cores in your CPU of course, but it's going up over time.

In a perfect works there would be some function like `File::is_seeking_going_to_be_slow`, and we could just ask the OS to tell us what to do. Unfortunately there's no such function.

The `--no-mmap` flag (or equivalently, taking input with `<` in the shell) is a fine workaround, but since mmapping provides a slight performance boost even in single-threaded mode, I think the best workaround is currently `--num-threads=1`.

**Comment by lulcat** [comment]
I guess I can confirm this. I solve this heuristically; however, so far I only did that 'solution' in my old C code. I haven't bothered yet doing it in my MT code, which indeed ironically also suffers from this and in this edge case is even slower than rust. It is a shame rust is being used in the first place, rather sub optimal. 

I have created a fat blakes binary which runs 2-3 times faster than your rust implementation. (~7 times faster than coreutil's single-threaded b2sum). I do suspect o'Connor is the rust fan of the team as it seems his projects are in rust. Such a shame too, I love bao, tao and lao but alas! all in rust, again, rather sub optimal but it is what it is.

I will release the code publically later but before release , I only am willing to share the binary. if you wanna check it does indeed run 2-3times faster than your rust oconnor: Shout and I will pass it on. This is not down to my code by the way, but down to your team's excellent code in assembly.  (So not the rust code).

[quote]
In a perfect works there would be some function like File::is_seeking_going_to_be_slow, and we could just ask the OS to tell us what to do. Unfortunately there's no such function.
 [/quote]

Have a look at what vmtouch and the underlying functions.. I am not sure if they can help in rust, especially since as you say, in MT, the seek pattern(s) will come at a cost. however, with a bit of heuristicvs one can get these times down in edge cases by a lot.
